### PR TITLE
Basic Constructs: Artificer

### PR DESCRIPTION
**Pull request by lizardqueenlexi**
## About The Pull Request

Really getting into the meat of the constructs now. Artificers have become basic mobs.

On the whole, this was a pretty rote conversion, with no significant gameplay changes other than the switch to using healing hands rather than a unique heal ability. The player experience as an artificer is more or less identical.

The _interesting_ part comes with the AI for the seldom-used "hostile" variant. Hostile artificers, being squishy and laughably weak, are now a dedicated "medic" role for constructs. They will perform triage, always seeking the most wounded construct (or shade!) to give healing to. They will not attack at all, but they _will_ flee with great speed if attacked and not busy healing. If they are healing another construct, they will remain even if they are beaten to death.

I've added some more AI functionality that may come in handy in the future, and done some refactoring to keep things from getting out of hand:
- A planning subtree for finding targets that will always select the most heavily wounded living target that the mob can see (or rather, the one with the least health). Useful again for medical triage, or for making a particularly cruel mob that always attacks whoever is easiest to kill. I plan to use this for NPC wraith constructs when I convert them.
- Targeting datums can now check a blackboard key to see if they should only target wounded mobs. This is particularly useful for "medic" type mobs such as this one.
- I've refactored the "minimum stat" behavior of targeting datums to be stored in a blackboard key. This removes the need to have unique subtypes for each different minimum stat we might want. Which... for the most part, weren't even used, leading to proliferation of several completely identical targeting datums in a bunch of different files. Hopefully this change will make things cleaner.

In addition, this PR fixes a pair of bugs from #78807 that I didn't catch:
- Healing constructs can now actually heal shades. Turns out I forgot to add the correct biotype.
- Healing hands, when set to print the target's remaining health, no longer does so as a visible message.

The one thing I didn't do that I kind of wanted to is make NPC artificers heal themselves when wounded and not busy doing something else, but it ended up being kind of annoying to make a mob willingly target itself. NPC artificers never had this behavior before, so I consider it okay, but maybe I'll circle back to it later.
## Why It's Good For The Game

Another basic conversion, another 5 items off the checklist. Very little should change in-game, though I think the new NPC AI could make for interesting challenges in ruins or bitrunning or something.
## Changelog
:cl:
refactor: Artificer constructs have been converted to the basic mob framework. This should change very little about them, but please report any bugs. NPC artificers are now smarter, and will focus on healing nearby wounded constructs - if you see them, take them out first!
/:cl:

### CODE REVIEW

File: code/datums/ai/basic_mobs/targetting_datums/basic_targetting_datum.dm [hunk 1]
```diff
@@ -17,8 +17,10 @@
 	var/check_factions_exactly = FALSE
 	/// Minimum status to attack living beings
 	var/stat_attack = CONSCIOUS
-	///Whether we care for seeing the target or not
+	/// Whether we care for seeing the target or not
 	var/ignore_sight = FALSE
+	/// Whether to target ONLY wounded targets
+	var/wounded_only = FALSE
```
Comment (Jacquerel) [inline code/datums/ai/basic_mobs/targetting_datums/basic_targetting_datum.dm:23]: I don't know if I love this being a flag on all instances of this behaviour, maybe we need a `check_valid_living` proc you can extend or something. 
Frankly a few of these should be on the blackboard too, ideally don't want to start proliferating subtypes with inheritance largely based on different combinations of booleans.

File: code/datums/ai/basic_mobs/targetting_datums/basic_targetting_datum.dm [hunk 2]
```diff
@@ -132,3 +136,12 @@
 
 /datum/targetting_datum/basic/attack_even_if_dead
 	stat_attack = DEAD
+
+/datum/targetting_datum/basic/attack_wounded_only
+	wounded_only = TRUE
+
+/// Intended mainly for "medic" mobs that have healing touch or some other support ability
+/datum/targetting_datum/basic/attack_wounded_only/same_faction
```
Comment (Jacquerel) [inline code/datums/ai/basic_mobs/targetting_datums/basic_targetting_datum.dm:144]: Here's some examples of subtypes which should preferably just be blackboard key behaviour (including the stat_attack ones you didn't add)

File: code/modules/antagonists/cult/runes.dm [hunk 3]
```diff
@@ -108,7 +108,7 @@ Runes can either be invoked by one's self or with many different cultists. Each
 
 /obj/effect/rune/attack_animal(mob/living/simple_animal/user, list/modifiers)
 	if(isshade(user) || isconstruct(user))
-		if(istype(user, /mob/living/simple_animal/hostile/construct/wraith/angelic) || istype(user, /mob/living/simple_animal/hostile/construct/juggernaut/angelic) || istype(user, /mob/living/simple_animal/hostile/construct/artificer/angelic))
+		if(istype(user, /mob/living/simple_animal/hostile/construct/wraith/angelic) || istype(user, /mob/living/simple_animal/hostile/construct/juggernaut/angelic) || istype(user, /mob/living/basic/construct/artificer/angelic))
```
Comment (Jacquerel) [inline code/modules/antagonists/cult/runes.dm:111]: This is an awful check which should be a trait, but you didn't add it so ignore it if you want.

File: code/modules/mob/living/basic/constructs/artificer.dm [hunk 4]
```diff
@@ -33,70 +29,33 @@
...
-
-/mob/living/simple_animal/hostile/construct/artificer/hostile //actually hostile, will move around, hit things, heal other constructs
-	AIStatus = AI_ON
-	environment_smash = ENVIRONMENT_SMASH_STRUCTURES //only token destruction, don't smash the cult wall NO STOP
+/// Hostile NPC version. Heals nearby constructs and cult structures, avoids targets that aren't extremely hurt.
+/mob/living/basic/construct/artificer/hostile
+	ai_controller = /datum/ai_controller/basic_controller/artificer
+	smashes_walls = FALSE
 
 /////////////////////////////Artificer-alts/////////////////////////
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/constructs/artificer.dm:47]: Minor but can we remove the big comment headers to just use `//` ?
It fucks with autodocs

File: code/modules/mob/living/basic/constructs/construct_ai.dm [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+		/datum/ai_planning_subtree/flee_target/from_flee_key,
+	)
+
+/// Targetting datum that will only allow mobs that constructs can heal.
+/datum/targetting_datum/basic/attack_wounded_only/same_faction/construct
+
+/datum/targetting_datum/basic/attack_wounded_only/same_faction/construct/can_attack(mob/living/living_mob, atom/the_target, vision_range, check_faction = TRUE)
+	. = ..()
+	if(isconstruct(the_target) || istype(the_target, /mob/living/simple_animal/shade))
+		return .
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/constructs/construct_ai.dm:23]:
```suggestion
	if(isconstruct(the_target) || istype(the_target, /mob/living/simple_animal/shade))
		return ..()
```
no point doing the super call if we know we're returning false

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by Yobrocharlie** [comment]
Good PR

**Comment by lizardqueenlexi** [comment]
Drafting because fucking with targetting datums has caused some problems I need to fix.

**Comment by san7890** [comment]
conflicts

**Pull request merged**